### PR TITLE
refactor: Optimisations fix

### PR DESCRIPTION
**Pull request by fix**
WIP

### CODE REVIEW

**Comment by alexbarnsley** [comment]
Your changes now make it so that it gets recent blocks from the database every time a block is downloaded, instead of the in-memory array.

**Comment by fix** [comment]
yes but there is no garantee about the state of this.recentBlockId
there is nowhere in the code when ids are removed when a block is reverted for instance

**Comment by fix** [comment]
this will provoke the network to fork hard

**Comment by alexbarnsley** [comment]
So why not add in a "force from database" and run that once all blocks are reverted. you complained about common blocks causing slowness in my PR, yet this will make it worse. just my $0.02.

**Comment by fix** [comment]
well i prefer slow and not forking than fast and forking

**Comment by fix** [comment]
also, i think these 'last blocks' should be managed centrally at the level of te state machine (together with other stateful data)

**Comment by fix** [comment]
Also essentially keeping last 100s full blocks in memory will help syncing fast, most peers will only download last 100 blocks (in worst case scenario when removing blocks)

File: packages/core-blockchain/lib/blockchain.js [hunk 1]
```diff
@@ -263,80 +264,81 @@ module.exports = class Blockchain {
...
     } else {
       logger.warn(`Block ${block.data.height.toLocaleString()} disregarded because verification failed :scroll:`)
-      callback()
+      return callback()
     }
   }
 
   /**
    * Process the given block.
+   * We should be sure this is fail safe (ie callback() is being called)
```
Comment (faustbrian) [inline packages/core-blockchain/lib/blockchain.js]: This is a `TODO:`.

File: packages/core-database-sequelize/lib/connection.js [hunk 2]
```diff
@@ -588,18 +588,14 @@ module.exports = class SequelizeConnection extends ConnectionInterface {
    * @return {[]String}
    */
   async getRecentBlockIds () {
-    if (!this.recentBlockIds.length) {
```
Comment (faustbrian) [inline packages/core-database-sequelize/lib/connection.js]: This is a big hit on performance as already tested in the past. Constantly hitting the blocks table for a read operation is resource intense and can easily result in a deadlock.

**Comment by fix** [comment]
adding an issue to track this down and clarify this

File: packages/core-blockchain/lib/blockchain.js (hunk 1 above)
Comment (fix) [reply packages/core-blockchain/lib/blockchain.js]: no i have made it failsafe, putting it as a reminder if somebody touch it

File: packages/core-blockchain/lib/blockchain.js [hunk 3]
```diff
@@ -263,80 +264,81 @@ module.exports = class Blockchain {
...
    * @return {(Function|void)}
    */
   async processBlock (block, callback) {
     if (!block.verification.verified) {
       logger.warn(`Block ${block.data.height.toLocaleString()} disregarded because verification failed :scroll:`)
-
       return callback()
     }
 
-    const state = stateMachine.state
```
Comment (faustbrian) [inline packages/core-blockchain/lib/blockchain.js]: Missing tests for the changes from here to `return this.dispatch('FORK')`.

File: packages/core-database-sequelize/lib/connection.js (hunk 2 above)
Comment (fix) [reply packages/core-database-sequelize/lib/connection.js]: see my comment for @alexbarnsley this.recentBlockIds as no garantee for being in sync with the state of the blockchain

File: packages/core-blockchain/lib/blockchain.js (hunk 1 above)
Comment (faustbrian) [reply packages/core-blockchain/lib/blockchain.js]: Then `NOTE:` to make it clearly visible that it should be read.

File: packages/core-database-sequelize/lib/connection.js (hunk 2 above)
Comment (faustbrian) [reply packages/core-database-sequelize/lib/connection.js]: Then `should be managed centrally at the level of the state machine` should be included in this PR as it is part of the same behaviour that is being modified.

File: packages/core-blockchain/lib/blockchain.js (hunk 3 above)
Comment (fix) [reply packages/core-blockchain/lib/blockchain.js]: i have just put the same logic to the calling function instead to be sure this is threadsafe.
IE moving logic from `acceptChainedBlock` to `processBlock`

File: packages/core-database-sequelize/lib/connection.js (hunk 2 above)
Comment (fix) [reply packages/core-database-sequelize/lib/connection.js]: this more complex than that, opening an issue for that, see again te discussion

File: packages/core-blockchain/lib/blockchain.js (hunk 1 above)
Comment (fix) [reply packages/core-blockchain/lib/blockchain.js]: OK

**Comment by faustbrian** [comment]
Travis is currently failing and in the next PR please respect the guidelines for commits https://docs.ark.io/developers/guidelines/git-commit-guidelines.html#types to make it easier to compile changelogs.

Also always make sure that PRs are squashed according to guidelines https://docs.ark.io/developers/guidelines/repository-management.html#squashing-pull-requests with a meaningful commit.

**Comment by fix** [comment]
https://github.com/ArkEcosystem/core/issues/818

**Comment by fix** [comment]
@faustbrian yo will do

**Pull request merged**